src/server/clank/backendAdb.ts

```diff
@@ -0,0 +1,153 @@
+    assert(status.toString() === 'OKAY');
+  }
+  await socket.write(encodeMessage(command));
+  const status = await socket.read(4);
+  assert(status.toString() === 'OKAY');
+  if (!command.startsWith('shell:')) {
+    const remainingLength = parseInt((await socket.read(4)).toString(), 16);
+    return (await socket.read(remainingLength)).toString();
+  }
+  return (await socket.readAll()).toString();
+}
+
+async function open(command: string, serial?: string): Promise<BufferedSocketWrapper> {
+  const socket = new BufferedSocketWrapper(net.createConnection({ port: 5037 }));
+  if (serial) {
+    await socket.write(encodeMessage(`host:transport:${serial}`));
+    const status = await socket.read(4);
+    assert(status.toString() === 'OKAY');
+  }
+  await socket.write(encodeMessage(command));
+  const status = await socket.read(4);
+  assert(status.toString() === 'OKAY');
+  return socket;
+}
+
+function encodeMessage(message: string): Buffer {
+  let lenHex = (message.length).toString(16);
+  lenHex = '0'.repeat(4 - lenHex.length) + lenHex;
+  return Buffer.from(lenHex + message);
+}
+
+class BufferedSocketWrapper extends EventEmitter implements SocketBackend {
+  private _socket: net.Socket;
+  private _buffer = Buffer.from([]);
+  private _isSocket = false;
+  private _notifyReader: (() => void) | undefined;
+  private _connectPromise: Promise<void>;
+  private _isClosed = false;
+
+  constructor(socket: net.Socket) {
+    super();
+    this._socket = socket;
+    this._connectPromise = new Promise(f => this._socket.on('connect', f));
+    this._socket.on('data', data => {
+      debug('pw:android:adb:data')(data.toString());
+      if (this._isSocket) {
+        this.emit('data', data);
+        return;
+      }
+      this._buffer = Buffer.concat([this._buffer, data]);
+      if (this._notifyReader)
+        this._notifyReader();
+    });
+    this._socket.on('close', () => {
+      this._isClosed = true;
+      if (this._notifyReader)
+        this._notifyReader();
+      this.emit('close');
+    });
+    this._socket.on('error', error => this.emit('error', error));
+  }
+
+  async write(data: Buffer) {
+    debug('pw:android:adb:send')(data.toString());
+    await this._connectPromise;
+    await new Promise(f => this._socket.write(data, f));
+  }
+
+  async close() {
+    debug('pw:android:adb')('Close');
+    this._socket.destroy();
+  }
+
+  async read(length: number): Promise<Buffer> {
+    await this._connectPromise;
+    assert(!this._isSocket, 'Can not read by length in socket mode');
+    while (this._buffer.length < length)
+      await new Promise(f => this._notifyReader = f);
+    const result = this._buffer.slice(0, length);
+    this._buffer = this._buffer.slice(length);
+    debug('pw:android:adb:recv')(result.toString());
+    return result;
+  }
+
+  async readAll(): Promise<Buffer> {
+    while (!this._isClosed)
+      await new Promise(f => this._notifyReader = f);
+    return this._buffer;
+  }
+
+  becomeSocket() {
```

Comment:
WDYT about external "become socket"? It could make things easier to understand.

```ts

// Pump
bufferedSocket.on('close', () => realSocket.emit('close'));
while (true)
  realSocket.emit('data', await bufferedSocket.readNext());

// readNext:
async readNext(): Promise<Buffer> {
  await new Promise(f => this._notifyReader = f);
  const buffer = this._buffer;
  this._buffer = Buffer.from([]);
  return buffer;
}

```

Reply:
Sorry, I didn't get it